### ISSUE TITLE
Adding a custom dataset with a predefined train/test split

### ISSUE THREAD

**Issue by btseytlin** [KevinMusgrave/powerful-benchmarker#63]
Hello, I am working on adding the Hotels-50k dataset for my project.
I have organized the images into the following structure:
```
root/
  - train/
     - label1/
        - image1.png
   - test/
     - label1/
        - image999.png
```

This dataset has a predefined train/test split. The splits also are not class disjoint.

How do I use this dataset with powerful benchmarker? I want to train and perform the 4-folds CV on the train split by the usual procedure, test on the test split. This is done so that I can compare with other works related to this dataset.

I was following [this reply](https://github.com/KevinMusgrave/powerful-benchmarker/issues/48#issuecomment-673125087), but currently I I am stuck at this point:
```
class Hotels50KDataset(Dataset):
    def __init__(self, root, transform=None, download=False):
        if download:
            retcode = download_hotels_50k(root)

        symlinks_dir = os.path.join(root, 'symlinks')
        symlinks_train_dir = os.path.join(symlinks_dir, 'train')
        symlinks_test_dir = os.path.join(symlinks_dir, 'test')
        
        # At this point I have image folders for train and test, but the class requires only one self.dataset. What now?
        self.dataset = datasets.ImageFolder(root, transform=transform)

        # these look useless, but are required by powerful-benchmarker
        self.labels = np.array([b for (a, b) in self.dataset.imgs])
        self.transform = transform 
 
    def __len__(self):
        return len(self.dataset)

    def __getitem__(self, idx):
        return self.dataset[idx]
```

Should I merge these two image folders together and use a split manager to separate train from test?

**Comment by KevinMusgrave** [comment]
Do you want the train/val splits to still be class disjoint, even though the trainval and test have overlapping classes?

**Comment by btseytlin** [comment]
> Do you want the train/val splits to still be class disjoint, even though the trainval and test have overlapping classes?

No, I want them to be as in the original dataset (not class disjoint). Though it would be good to know how to make them class disjoint too.

**Comment by KevinMusgrave** [comment]
Alright, here's my shitty solution for now. The train/val splits will not be class disjoint. I actually don't have a good way of making them class disjoint, without also making them class disjoint from the test set.

Here is a test script you can run separately from the benchmarker. You need to replace "CustomCUB200" with your own dataset (don't extend CUB200, just replace the whole thing with your own dataset). Then you need to implement ```get_split_indices```, which returns the indices of the test set, within the entire unsplit dataset. So you'll need to make your ```self.dataset``` consist of the the training set concatenated with the test set.

```python
from powerful_benchmarker.datasets import CUB200
from powerful_benchmarker.split_managers import ClassDisjointSplitManager, BaseSplitManager, IndexSplitManager
from torchvision import transforms
import torch
import numpy as np

class CustomSplitManager(BaseSplitManager):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.split_names = ["test"]

    def _create_split_schemes(self, datasets):
        output = {}
        for transform_type, v1 in datasets.items():
            output[transform_type] = {}
            for split_name, v2 in v1.items():
                indices = v2.get_split_indices(split_name)
                if indices is not None:
                    output[transform_type][split_name] = torch.utils.data.Subset(v2, indices)
                else:
                    output[transform_type][split_name] = v2
        return {self.get_split_scheme_name(0): output}

    def get_test_set_name(self):
        return 'UsingOriginalTest'

    def get_base_split_scheme_name(self):
        return self.get_test_set_name()

    def get_split_scheme_name(self, partition):
        return self.get_base_split_scheme_name()

    def split_assertions(self):
        pass


class CustomCUB200(CUB200):
    def get_split_indices(self, split_name):
        if split_name == "test":
            return np.random.choice(np.arange(10000), size=6666, replace=False)
        return None


eval_transform = transforms.Compose([
                    transforms.Resize(256),
                    transforms.CenterCrop(227),
                    transforms.ToTensor(),
                    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
                ])

dataset_root = "/scratch"
the_dataset = CustomCUB200

# first level of dictionary is transform type
# second level consists of separate dataset objects for each split
datasets = {"eval": {"train": the_dataset(dataset_root, eval_transform),
                    "val": the_dataset(dataset_root, eval_transform),
                    "test": the_dataset(dataset_root, eval_transform)}
            }

helper_split_manager = CustomSplitManager()
split_manager = IndexSplitManager(num_training_partitions=4, num_training_sets=4, helper_split_manager=helper_split_manager)

split_manager.create_split_schemes(datasets)

test_set = split_manager.get_dataset("eval", "test")

for name in split_manager.split_scheme_names:
    split_manager.set_curr_split_scheme(name)
    print("CURR SPLIT SCHEME =", split_manager.curr_split_scheme_name)
    print("train len", len(split_manager.get_dataset("eval", "train")))
    print("val len", len(split_manager.get_dataset("eval", "val")))
    print("test len", len(split_manager.get_dataset("eval", "test")))

    print("train labels", split_manager.get_label_set("eval", "train"))
    print("val labels", split_manager.get_label_set("eval", "val"))
    print("test labels", split_manager.get_label_set("eval", "test"))
```

Once you have the above working, register the custom split manager, and use these flags in your script:
```
--split_manager~SWAP~1 {IndexSplitManager: {}} \
--split_manager~APPLY~2 {helper_split_manager: {CustomSplitManager: {}}} \
```

**Comment by KevinMusgrave** [comment]
This might work inside your dataset class:

```python
def __init__(...):
  original_test_dataset = datasets.ImageFolder(...)
  original_train_dataset = datasets.ImageFolder(...) 
  self.dataset = torch.utils.data.ConcatDataset([original_test_dataset, original_train_dataset])
  self.test_set_indices = np.arange(len(original_test_dataset))

def get_split_indices(self, split_name):
  if split_name == "test":
    return self.test_set_indices
  return None
```

**Comment by btseytlin** [comment]
Thanks, I will try that!
